pkg/runtime/runtime.go

```diff
@@ -36,13 +53,42 @@ type Info struct {
 	Trainer
 	// Scheduler parameters to add to the RuntimeJobTemplate.
 	*Scheduler
+	// TemplateSpec is TrainingRuntime Template object.
+	// ObjApply podSpecs and this PodSets should be kept in sync by info.SyncPodSetsToTemplateSpec().
+	TemplateSpec TemplateSpec
 }
 
 type RuntimePolicy struct {
 	MLPolicy       *trainer.MLPolicy
 	PodGroupPolicy *trainer.PodGroupPolicy
 }
 
+type TemplateSpec struct {
+	// ObjApply is ApplyConfiguration for the TrainingRuntimes Template field.
+	ObjApply any
+	// PodSets is a set of Pod extracted from ObjApply.
+	PodSets []PodSet
+}
+
+type PodSet struct {
+	Name string
+	// If Name is trainer-node, CountForNonTrainer is null.
+	// For Trainer, PodSet Count should be stored in Info.RuntimePolicy.MLPolicy.NumNodes.
+	CountForNonTrainer *int32
```

Comment:
Why do we need it if we are planning to support more than 1 Job replica for other other Jobs (e.g. Initializer, Launcher).

Reply:
We actually support it here. The key point is that we have 3 type of replication, `numNodes`, `batchv1.Job.parallelism`, and `JobSet.replicatedJob.Replica`.

This `CountForNonTrainer` corresponds to `batchv1.Job.parallelism`.

Reply:
Do we want to support `batchv1.Job.parallelism != 1` for non Trainer Node jobs ?

Reply:
They can specify arbitrary roles except for trainer-node, initializer, and launcher. And it would be better to just propagate the parallelism parameters to JobSet.

Reply:
I see makes sense.
I am wondering whether we should consolidate NumNodes to the PodCount as well ?
Should we just override this value in the MLPolicy plugins based on `trainJob.spec.numNodes` and `trainingRuntime.spec.mlPolicy.numNodes`

Reply:
> I am wondering whether we should consolidate NumNodes to the PodCount as well ?
Should we just override this value in the MLPolicy plugins based on trainJob.spec.numNodes and trainingRuntime.spec.mlPolicy.numNodes

Yeah, I was thinking the same thing. However, if we want to do it, we need to change Info.RuntimePolicy data structure, and it will affect all plugin implementation. So, let me revisit if we should do it after https://github.com/kubeflow/trainer/issues/2495

If we make a lot of data structure changes in a single scope, that causes bugs, IMO.

Reply:
Sounds great! It would be nice to make a note about it in the #2495

Reply:
Yeah sure

Reply:
Done.

> After the above migration, we want to consider dropping NumNodes from Info.RuntimePolicy.MLPolicy and then fully rely on the PodSet data structure in case of Trainer as well.